node/cli/src/command.rs

```diff
@@ -0,0 +1,797 @@
+	fn impl_name() -> String {
+		#[cfg(feature = "with-moonbeam-runtime")]
+		return "Moonbeam Parachain Collator".into();
+		#[cfg(feature = "with-moonriver-runtime")]
+		return "Moonriver Parachain Collator".into();
+		#[cfg(feature = "with-moonbase-runtime")]
+		return "Moonbase Parachain Collator".into();
+	}
+
+	fn impl_version() -> String {
+		env!("SUBSTRATE_CLI_IMPL_VERSION").into()
+	}
+
+	fn description() -> String {
+		"Moonbeam Parachain Collator\n\nThe command-line arguments provided first will be \
+		passed to the parachain node, while the arguments provided after -- will be passed \
+		to the relaychain node.\n\n\
+		parachain-collator [parachain-args] -- [relaychain-args]"
+			.into()
+	}
+
+	fn author() -> String {
+		env!("CARGO_PKG_AUTHORS").into()
+	}
+
+	fn support_url() -> String {
+		"https://github.com/PureStake/moonbeam/issues/new".into()
+	}
+
+	fn copyright_start_year() -> i32 {
+		2019
+	}
+
+	fn load_spec(&self, id: &str) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
+		match id {
+			"moonbase_alpha_relay" => Ok(Box::new(RococoChainSpec::from_json_bytes(
+				&include_bytes!("../../../specs/alphanet/rococo-embedded-specs-v7.json")[..],
+			)?)),
+			"moonbase_stage_relay" => Ok(Box::new(RococoChainSpec::from_json_bytes(
+				&include_bytes!("../../../specs/stagenet/rococo-embedded-specs-v7.json")[..],
+			)?)),
+			// If we are not using a moonbeam-centric pre-baked relay spec, then fall back to the
+			// Polkadot service to interpret the id.
+			_ => polkadot_cli::Cli::from_iter([RelayChainCli::executable_name()].iter())
+				.load_spec(id),
+		}
+	}
+
+	fn native_runtime_version(chain_spec: &Box<dyn ChainSpec>) -> &'static RuntimeVersion {
+		polkadot_cli::Cli::native_runtime_version(chain_spec)
+	}
+}
+
+#[allow(clippy::borrowed_box)]
+fn extract_genesis_wasm(chain_spec: &Box<dyn sc_service::ChainSpec>) -> Result<Vec<u8>> {
+	let mut storage = chain_spec.build_storage()?;
+
+	storage
+		.top
+		.remove(sp_core::storage::well_known_keys::CODE)
+		.ok_or_else(|| "Could not find wasm file in genesis state!".into())
+}
+
+/// Parse command line arguments into service configuration.
+pub fn run() -> Result<()> {
+	let cli = Cli::from_args();
+	let author_id: Option<H160> = cli.run.author_id;
+	match &cli.subcommand {
+		Some(Subcommand::BuildSpec(params)) => {
+			let runner = cli.create_runner(&params.base)?;
+			runner.sync_run(|config| {
+				if params.mnemonic.is_some() || params.accounts.is_some() {
+					if config.chain_spec.is_moonbeam() {
+						// TODO-multiple-runtimes
+						#[cfg(feature = "with-moonbeam-runtime")]
+						{
+							params.base.run(
+								Box::new(chain_spec::moonbeam::development_chain_spec(
+									params.mnemonic.clone(),
+									params.accounts,
+								)),
+								config.network,
+							)
+						}
+						#[cfg(not(feature = "with-moonbeam-runtime"))]
+						return Err(MISSING_MOONBEAM.into());
+					} else if config.chain_spec.is_moonriver() {
```

Comment:
It gets a little hard to read. I don't immediately see a way to make it more than marginally better. I guess we should really just have good tests to make sure we've defined the expected behavior and achieved it.